go-controller/pkg/ovn/external_ids_syncer/acl/acl_sync.go

```diff
@@ -67,75 +69,110 @@ func (syncer *aclSyncer) SyncACLs(existingNodes *v1.NodeList) error {
-	}
-	// 2. Check to-be-updated ACLs don't have the same PrimaryID between themselves and with the existingACLPrimaryIDs
-	uniquePrimaryIDACLs := []*nbdb.ACL{}
-	for _, acl := range updatedACLs {
-		primaryID := acl.ExternalIDs[libovsdbops.PrimaryIDKey.String()]
-		if existingACLPrimaryIDs.Has(primaryID) {
-			// don't update that acl, otherwise 2 ACLs with the same primary ID will be in the db
-			klog.Warningf("Skip updating ACL %+v to the new ExternalIDs, since there is another ACL with the same primary ID", acl)
-		} else {
-			existingACLPrimaryIDs.Insert(primaryID)
-			uniquePrimaryIDACLs = append(uniquePrimaryIDACLs, acl)
+		// 2. Check to-be-updated ACLs don't have the same PrimaryID between themselves and with the existingACLPrimaryIDs
+		uniquePrimaryIDACLs := []*nbdb.ACL{}
+		for _, acl := range updatedACLs {
+			primaryID := acl.ExternalIDs[libovsdbops.PrimaryIDKey.String()]
+			if existingACLPrimaryIDs.Has(primaryID) {
+				// don't update that acl, otherwise 2 ACLs with the same primary ID will be in the db
+				klog.Warningf("Skip updating ACL %+v to the new ExternalIDs, since there is another ACL with the same primary ID", acl)
+			} else {
+				existingACLPrimaryIDs.Insert(primaryID)
+				uniquePrimaryIDACLs = append(uniquePrimaryIDACLs, acl)
+			}
+		}
+
+		// update acls with new ExternalIDs
+		err = batching.Batch[*nbdb.ACL](syncer.txnBatchSize, uniquePrimaryIDACLs, func(batchACLs []*nbdb.ACL) error {
+			return libovsdbops.CreateOrUpdateACLs(syncer.nbClient, batchACLs...)
+		})
+		if err != nil {
+			return fmt.Errorf("cannot update stale ACLs: %v", err)
 		}
 	}
 
-	// update acls with new ExternalIDs
-	err = batching.Batch[*nbdb.ACL](syncer.txnBatchSize, uniquePrimaryIDACLs, func(batchACLs []*nbdb.ACL) error {
-		return libovsdbops.CreateOrUpdateACLs(syncer.nbClient, batchACLs...)
-	})
+	// Once all the staleACLs are deleted and the externalIDs have been updated (externalIDs update should be a one-time
+	// upgrade operation), let us now update the tier's of all existing ACLs to types.DefaultACLTier. During upgrades after
+	// the OVN schema changes are applied, the nbdb.ACL.Tier column will be added and every row will be updated to 0 by
+	// default (types.PlaceHolderACLTier). For all features using ACLs (egressFirewall, NetworkPolicy, NodeACLs) we want to
+	// move them to Tier2. We need to do this in reverse order of ACL priority to avoid network traffic disruption during
+	// upgrades window (if not done according to priorities we might end up in a window where the ACL with priority 1000
+	// for default deny is in tier0 while 1001 ACL for allow-ing traffic is in tier2 for a given namespace network policy).
+	// NOTE: This is a one-time operation as no ACLs should ever be created in types.PlaceHolderACLTier moving forward.
+	// Fetch all ACLs in types.PlaceHolderACLTier (Tier0); update their Tier to 2 and batch the ACL update.
+	klog.Info("Updating Tier of existing ACLs...")
```

Comment:
Any chance to have a test for this?

Reply:
so we have tests in the `go-controller/pkg/ovn/external_ids_syncer/acl/acl_sync_test.go`; specifically here: https://github.com/ovn-org/ovn-kubernetes/pull/3645/commits/b53b9f752db4e530c015ae135ce2d59de0d31930#diff-fbf87e9a26d9865f6ce3881764044151ed71700db9c0440b1c527acba89024f9R451 and https://github.com/ovn-org/ovn-kubernetes/pull/3645/commits/b53b9f752db4e530c015ae135ce2d59de0d31930#diff-fbf87e9a26d9865f6ce3881764044151ed71700db9c0440b1c527acba89024f9R593 -> those parts are starting the tests with Placeholder ACLs and then moving them to Tier2. That tests this specific code path...
Did you have something else in mind?

Also I tried to write the test to check if lower priorities are moved before higher, but I didn't have bright ideas there..